core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BpmBillingProjectLifecycle.scala

```diff
@@ -97,26 +104,106 @@ class BpmBillingProjectLifecycle(
             RawlsErrorReport(StatusCode.int2StatusCode(errorReport.getStatusCode), errorReport.getMessage)
           )
         case None =>
+          landingZoneId = Some(landingZoneResponse.getLandingZoneId)
           logger.info(
-            s"Creating BPM-backed billing project ${projectName}, initiated creation of landing zone ${landingZoneResponse.getLandingZoneId} with jobId ${landingZoneResponse.getJobReport.getId}"
+            s"Creating BPM-backed billing project ${projectName.value}, initiated creation of landing zone ${landingZoneId.get} with jobId ${landingZoneResponse.getJobReport.getId}"
           )
       }
-      for {
-        _ <- billingRepository.updateLandingZoneId(createProjectRequest.projectName,
-                                                   landingZoneResponse.getLandingZoneId
-        )
+      (for {
+        _ <- billingRepository.updateLandingZoneId(createProjectRequest.projectName, landingZoneId.get)
         _ <- resourceMonitorRecordDao.create(
           UUID.fromString(landingZoneResponse.getJobReport.getId),
           JobType.AzureLandingZoneResult,
-          projectName
+          projectName.value
         )
         _ <- billingRepository.setBillingProfileId(createProjectRequest.projectName, profileModel.getId)
-      } yield CreationStatuses.CreatingLandingZone
+      } yield CreationStatuses.CreatingLandingZone).recoverWith { case t: Throwable =>
```

Comment:
I was surprised that the catch down at line 126 doesn't cover an error thrown within the for comprehension. Is there a cleaner way for me to write this code to avoid the duplicated cleanup logic?

Reply:
Yeah and it's not a huge rewrite! When it comes to error handling and cleaning up resources on error, lexical scoping is your friend (though it does lead to some indentation). This allows you to register separate error handlers to clean up errors as the stack gets popped. Doing so can really clean up your error handling logic as you don't have to worry about Option types - you know the resource exists. The general pattern is as follows: 
```scala
createResourceA.flatMap { resourceA => 
   createResourceB(A).flatMap { resourceB
      useResourceB(resourceB).recoverWith { t: Throwable =>
         destroyB(resourceB) >> failed(t)
      }
   }.recoverWith { t: Throwable => 
     destroyA(resourceA) >> failed(t)
   }
}
```

Applying this pattern to the function above becomes: 
```scala
    override def postCreationSteps(createProjectRequest: CreateRawlsV2BillingProjectFullRequest,
                                 config: MultiCloudWorkspaceConfig,
                                 ctx: RawlsRequestContext
  ): Future[CreationStatus] = {
    val projectName = createProjectRequest.projectName

    def createBillingProfile: Future[ProfileModel] =
      Future(blocking {
        val profileModel = billingProfileManagerDAO.createBillingProfile(
          projectName.value,
          createProjectRequest.billingInfo,
          ctx
        )
        logger.info(s"Created Billing Profile with ID ${profileModel.getId} for Billing Project $projectName.")
        profileModel
      })

    // This starts a landing zone creation job. There is a separate monitor that polls to see when it
    // completes and then updates the billing project status accordingly.
    def createLandingZone(profileModel: ProfileModel): Future[CreateLandingZoneResult] =
      Future(blocking {
        workspaceManagerDAO.createLandingZone(
          config.azureConfig.get.landingZoneDefinition,
          config.azureConfig.get.landingZoneVersion,
          profileModel.getId,
          ctx
        )
      })

    createBillingProfile.flatMap { profileModel =>
      createLandingZone(profileModel)
        .flatMap { landingZone =>
          (for {
            _ <- Option(landingZone.getErrorReport).traverse_ { errorReport =>
              Future.failed(
                new LandingZoneCreationException(
                  RawlsErrorReport(StatusCode.int2StatusCode(errorReport.getStatusCode), errorReport.getMessage)
                )
              )
            }
            _ <- billingRepository.updateLandingZoneId(createProjectRequest.projectName, landingZone.getLandingZoneId)
            _ <- resourceMonitorRecordDao.create(
              UUID.fromString(landingZone.getJobReport.getId),
              JobType.AzureLandingZoneResult,
              projectName.value
            )
            _ <- billingRepository.setBillingProfileId(createProjectRequest.projectName, profileModel.getId)
          } yield CreationStatuses.CreatingLandingZone).recoverWith { case t: Throwable =>
            // no Option here
            cleanupLandingZone(landingZone.getLandingZoneId, projectName, ctx) >> Future.failed(t)
          }
        }
        .recoverWith { case t: Throwable =>
          // no Option here
          cleanupBillingProfile(profileModel.getId, projectName, ctx) >> Future.failed(t)
        }
    }
  }
```